File: CartQuestiOS/CartQuestiOS/Views/Shop/ShopViewModel.swift
import CoreLocation
import Foundation
import Observation

@Observable
class ShopViewModel {

    // MARK: - Nested Types

    struct ProductResult: Identifiable {
        let product: KrogerProduct
        let isAvailable: Bool

        var id: String { product.productId }

        var price: Double? {
            product.items?.first?.price?.regular
        }

        var imageUrl: String? {
            product.bestImageUrl
        }
    }

    // MARK: - State

    var searchQuery: String = ""
    var searchResults: [ProductResult] = []
    var isSearching: Bool = false
    var hasSearched: Bool = false

    var cart: Cart = Cart()
    var isSaving: Bool = false

    var nearbyLocationIds: [String] = []
    var locationError: String? = nil

    // MARK: - Private

    private let krogerService: KrogerService
    private let cartRepository: CartRepository
    private let locationService: LocationService

    private var searchTask: Task<Void, Never>?
    private var saveTask: Task<Void, Never>?

    // MARK: - Init

    init(
        krogerService: KrogerService = KrogerService(
            clientId: Bundle.main.infoDictionary?["KROGER_CLIENT_ID"] as? String ?? "",
            clientSecret: Bundle.main.infoDictionary?["KROGER_CLIENT_SECRET"] as? String ?? ""
        ),
        cartRepository: CartRepository = CartRepository(),
        locationService: LocationService = LocationService()
    ) {
        self.krogerService = krogerService
        self.cartRepository = cartRepository
        self.locationService = locationService
        Task { await loadActiveCart() }
        Task { await fetchNearbyLocation() }
    }

    private init(skipInit: Bool) {
        self.krogerService = KrogerService(clientId: "", clientSecret: "")
        self.cartRepository = CartRepository()
        self.locationService = LocationService()
    }

    // MARK: - Location

    private func fetchNearbyLocation() async {
        do {
            let coordinate = try await locationService.getCurrentLocation()
            let stores = try await krogerService.searchLocations(
                lat: coordinate.latitude,
                lon: coordinate.longitude,
                radiusInMiles: 10,
                limit: 50
            )
            nearbyLocationIds = stores.map { $0.locationId }
        } catch {
            locationError = error.localizedDescription
        }
    }

    // MARK: - Cart Loading

    private func loadActiveCart() async {
        do {
            if let activeCart = try await cartRepository.getActiveCart() {
                cart = activeCart
            }
        } catch {
            print("Failed to load active cart: \(error)")
        }
    }

    // MARK: - Search

    func search() {
        let query = searchQuery.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !query.isEmpty else { return }
        searchTask?.cancel()
        searchTask = Task { await performSearch(query: query) }
    }

    private func performSearch(query: String) async {
        isSearching = true
        defer {
            isSearching = false
            hasSearched = true
        }

        let locationIds = nearbyLocationIds
        guard !locationIds.isEmpty else {
            // No stores found — search without location filter as fallback
            do {
                let products = try await krogerService.searchProducts(term: query, limit: 50)
                guard !Task.isCancelled else { return }
                searchResults = products.compactMap { product in
                    let inStore = product.items?.first?.fulfillment?.inStore ?? false
                    guard inStore else { return nil }
                    return ProductResult(product: product, isAvailable: true)
                }
            } catch {
                guard !Task.isCancelled else { return }
                print("Search failed: \(error)")
                searchResults = []
            }
            return
        }

        // Search all nearby stores in parallel
        let results = await withTaskGroup(of: [KrogerProduct].self) { group in
            for locationId in locationIds {
                group.addTask { [krogerService] in
                    (try? await krogerService.searchProducts(
                        term: query,
                        locationId: locationId,
                        limit: 50
                    )) ?? []
                }
            }
            var all: [KrogerProduct] = []
            for await batch in group {
                all.append(contentsOf: batch)
            }
            return all
        }

        guard !Task.isCancelled else { return }

        // Deduplicate by product ID, keeping the first occurrence
        var seen = Set<String>()
        searchResults = results.compactMap { product in
            let inStore = product.items?.first?.fulfillment?.inStore ?? false
            guard inStore, seen.insert(product.productId).inserted else { return nil }
            return ProductResult(product: product, isAvailable: true)
        }
    }

    func searchProducts(query: String) async -> [ProductResult] {
        let trimmed = query.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return [] }

        let locationIds = nearbyLocationIds
        if locationIds.isEmpty {
            do {
                let products = try await krogerService.searchProducts(term: trimmed, limit: 50)
                return products.compactMap { product in
                    let inStore = product.items?.first?.fulfillment?.inStore ?? false
                    guard inStore else { return nil }
                    return ProductResult(product: product, isAvailable: true)
                }
            } catch {
                return []
            }
        }

        let results = await withTaskGroup(of: [KrogerProduct].self) { group in
            for locationId in locationIds {
                group.addTask { [krogerService] in
                    (try? await krogerService.searchProducts(term: trimmed, locationId: locationId, limit: 50)) ?? []
                }
            }
            var all: [KrogerProduct] = []
            for await batch in group { all.append(contentsOf: batch) }
            return all
        }

        var seen = Set<String>()
        return results.compactMap { product in
            let inStore = product.items?.first?.fulfillment?.inStore ?? false
            guard inStore, seen.insert(product.productId).inserted else { return nil }
            return ProductResult(product: product, isAvailable: true)
        }
    }

    func clearSearch() {
        searchQuery = ""
        searchResults = []
        hasSearched = false
        searchTask?.cancel()
    }

    // MARK: - Cart Queries

    func cartQuantity(for productId: String) -> Int {
        cart.items.first(where: { $0.productId == productId })?.quantity ?? 0
    }

    // MARK: - Cart Mutations

    func addToCart(product: KrogerProduct) {
        let imageUrl = product.bestImageUrl ?? ""
        let item = CartItem(
            productId: product.productId,
            name: product.description,
            brand: product.brand ?? "",
            imageUrl: imageUrl,
            quantity: 1,
            substitutes: []
        )
        cart.items.append(item)
        debounceSave()
    }

    func incrementQuantity(productId: String) {
        guard let index = cart.items.firstIndex(where: { $0.productId == productId }) else { return }
        cart.items[index].quantity += 1
        debounceSave()
    }

    func decrementQuantity(productId: String) {
        guard let index = cart.items.firstIndex(where: { $0.productId == productId }) else { return }
        if cart.items[index].quantity <= 1 {
            cart.items.remove(at: index)
        } else {
            cart.items[index].quantity -= 1
        }
        debounceSave()
    }

    func removeFromCart(at index: Int) {
        guard cart.items.indices.contains(index) else { return }
        cart.items.remove(at: index)
        debounceSave()
    }

    // MARK: - Substitutes

    func addSubstitute(to cartItemIndex: Int, product: KrogerProduct) {
        guard cart.items.indices.contains(cartItemIndex) else { return }
        let sub = Substitute(
            productId: product.productId,
            name: product.description,
            brand: product.brand ?? ""
        )
        // Don't add duplicate substitutes
        guard !cart.items[cartItemIndex].substitutes.contains(where: { $0.productId == sub.productId }) else { return }
        cart.items[cartItemIndex].substitutes.append(sub)
        debounceSave()
    }

    func removeSubstitute(from cartItemIndex: Int, substituteIndex: Int) {
        guard cart.items.indices.contains(cartItemIndex),
              cart.items[cartItemIndex].substitutes.indices.contains(substituteIndex) else { return }
        cart.items[cartItemIndex].substitutes.remove(at: substituteIndex)
        debounceSave()
    }

    var tripJustCompleted: Bool = false

    func clearCart() {
        saveTask?.cancel()
        cart = Cart()
        searchQuery = ""
        searchResults = []
        hasSearched = false
    }

    func updateQuantity(at index: Int, quantity: Int) {
        guard cart.items.indices.contains(index) else { return }
        if quantity < 1 {
            cart.items.remove(at: index)
        } else {
            cart.items[index].quantity = quantity
        }
        debounceSave()
    }

    // MARK: - Persistence

    func saveCartNow() async {
        saveTask?.cancel()
        saveTask = nil
        await saveCart()
    }

    private func debounceSave() {
        saveTask?.cancel()
        saveTask = Task {
            try? await Task.sleep(for: .milliseconds(800))
            guard !Task.isCancelled else { return }
            await saveCart()
        }
    }

    private func saveCart() async {
        isSaving = true
        defer { isSaving = false }
        do {
            cart = try await cartRepository.saveCart(cart)
        } catch {
            print("Failed to save cart: \(error)")
        }
    }
}

// MARK: - Preview Helper

extension ShopViewModel {
    static var preview: ShopViewModel {
        let vm = ShopViewModel(skipInit: true)
        vm.searchResults = [
            ProductResult(
                product: KrogerProduct(productId: "001", description: "Organic Strawberries, 1 Lb", brand: "Simple Truth", images: [], items: [KrogerItemPrice(price: KrogerPrice(regular: 3.99, promo: 3.99), fulfillment: KrogerFulfillment(inStore: true))]),
                isAvailable: true
            ),
            ProductResult(
                product: KrogerProduct(productId: "002", description: "Whole Milk, 1 Gallon", brand: "Kroger", images: [], items: [KrogerItemPrice(price: KrogerPrice(regular: 4.29, promo: 4.29), fulfillment: KrogerFulfillment(inStore: true))]),
                isAvailable: true
            )
        ]
        vm.cart = Cart(items: [CartItem(productId: "001", name: "Organic Strawberries, 1 Lb", brand: "Simple Truth", imageUrl: "", quantity: 2, substitutes: [])])
        return vm
    }
}

File: CartQuestAndroid/app/src/main/java/com/amoghghadge/cartquestandroid/ui/shop/ShopViewModel.kt
package com.amoghghadge.cartquestandroid.ui.shop

import android.annotation.SuppressLint
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.amoghghadge.cartquestandroid.BuildConfig
import com.amoghghadge.cartquestandroid.data.model.Cart
import com.amoghghadge.cartquestandroid.data.model.CartItem
import com.amoghghadge.cartquestandroid.data.model.KrogerProduct
import com.amoghghadge.cartquestandroid.data.remote.KrogerApiService
import com.amoghghadge.cartquestandroid.data.remote.KrogerAuthManager
import com.amoghghadge.cartquestandroid.data.repository.CartRepository
import com.amoghghadge.cartquestandroid.service.LocationService
import kotlinx.coroutines.Job
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

data class ProductResult(
    val product: KrogerProduct,
    val isAvailable: Boolean
) {
    val price: Double? get() = product.items?.firstOrNull()?.price?.regular
    val imageUrl: String? get() = product.bestImageUrl
}

data class ShopUiState(
    val searchQuery: String = "",
    val searchResults: List<ProductResult> = emptyList(),
    val isSearching: Boolean = false,
    val hasSearched: Boolean = false,
    val cart: Cart = Cart(),
    val isSaving: Boolean = false,
    val nearbyLocationIds: List<String> = emptyList(),
    val locationError: String? = null
)

class ShopViewModel : ViewModel() {

    private val _uiState = MutableStateFlow(ShopUiState())
    val uiState: StateFlow<ShopUiState> = _uiState.asStateFlow()

    private val cartRepository = CartRepository()
    private val krogerApiService = KrogerApiService.create()
    private val krogerAuthManager = KrogerAuthManager(
        clientId = BuildConfig.KROGER_CLIENT_ID,
        clientSecret = BuildConfig.KROGER_CLIENT_SECRET
    )

    @SuppressLint("StaticFieldLeak")
    private var locationService: LocationService? = null

    private var searchJob: Job? = null
    private var saveJob: Job? = null

    init {
        viewModelScope.launch {
            try {
                val activeCart = cartRepository.getActiveCart()
                if (activeCart != null) {
                    _uiState.update { it.copy(cart = activeCart) }
                }
            } catch (_: Exception) {
                // No active cart found, keep default empty cart
            }
        }
    }

    fun initLocation(locationService: LocationService) {
        this.locationService = locationService
        viewModelScope.launch {
            try {
                val latLng = locationService.getCurrentLocation()
                val token = krogerAuthManager.getToken()
                val response = krogerApiService.searchLocations(
                    auth = "Bearer $token",
                    lat = latLng.latitude,
                    lon = latLng.longitude
                )
                val locationIds = response.data.map { it.locationId }
                _uiState.update { it.copy(nearbyLocationIds = locationIds) }
            } catch (e: Exception) {
                _uiState.update { it.copy(locationError = e.message) }
            }
        }
    }

    fun updateSearchQuery(query: String) {
        _uiState.update { it.copy(searchQuery = query) }
        if (query.isBlank()) {
            searchJob?.cancel()
            _uiState.update { it.copy(searchResults = emptyList(), isSearching = false) }
        }
    }

    fun search() {
        val query = _uiState.value.searchQuery.trim()
        if (query.isBlank()) return
        searchJob?.cancel()
        searchJob = viewModelScope.launch {
            _uiState.update { it.copy(isSearching = true) }
            try {
                val token = krogerAuthManager.getToken()
                val locationIds = _uiState.value.nearbyLocationIds

                val allProducts = if (locationIds.isEmpty()) {
                    // Fallback: search without location filter
                    val response = krogerApiService.searchProducts(
                        auth = "Bearer $token",
                        term = query
                    )
                    response.data
                } else {
                    // Search all nearby stores in parallel
                    locationIds.map { locationId ->
                        async {
                            try {
                                krogerApiService.searchProducts(
                                    auth = "Bearer $token",
                                    term = query,
                                    locationId = locationId
                                ).data
                            } catch (_: Exception) {
                                emptyList()
                            }
                        }
                    }.awaitAll().flatten()
                }

                // Deduplicate by product ID, keep first occurrence
                val seen = mutableSetOf<String>()
                val results = allProducts.mapNotNull { product ->
                    val inStore = product.items?.firstOrNull()?.fulfillment?.inStore == true
                    if (inStore && seen.add(product.productId)) {
                        ProductResult(product = product, isAvailable = true)
                    } else null
                }
                _uiState.update { it.copy(searchResults = results, isSearching = false, hasSearched = true) }
            } catch (_: Exception) {
                _uiState.update { it.copy(isSearching = false, hasSearched = true) }
            }
        }
    }

    suspend fun searchProducts(query: String): List<ProductResult> {
        val trimmed = query.trim()
        if (trimmed.isBlank()) return emptyList()
        return try {
            val token = krogerAuthManager.getToken()
            val locationIds = _uiState.value.nearbyLocationIds
            val allProducts = if (locationIds.isEmpty()) {
                krogerApiService.searchProducts(auth = "Bearer $token", term = trimmed).data
            } else {
                locationIds.map { locationId ->
                    viewModelScope.async {
                        try {
                            krogerApiService.searchProducts(auth = "Bearer $token", term = trimmed, locationId = locationId).data
                        } catch (_: Exception) { emptyList() }
                    }
                }.awaitAll().flatten()
            }
            val seen = mutableSetOf<String>()
            allProducts.mapNotNull { product ->
                val inStore = product.items?.firstOrNull()?.fulfillment?.inStore == true
                if (inStore && seen.add(product.productId)) ProductResult(product, true) else null
            }
        } catch (_: Exception) { emptyList() }
    }

    fun clearSearch() {
        searchJob?.cancel()
        _uiState.update { it.copy(searchQuery = "", searchResults = emptyList(), hasSearched = false) }
    }

    fun cartQuantity(productId: String): Int {
        return _uiState.value.cart.items
            .firstOrNull { it.productId == productId }
            ?.quantity ?: 0
    }

    fun addToCart(product: KrogerProduct) {
        val imageUrl = product.bestImageUrl.orEmpty()
        val newItem = CartItem(
            productId = product.productId,
            name = product.description,
            brand = product.brand.orEmpty(),
            imageUrl = imageUrl,
            quantity = 1
        )
        _uiState.update { state ->
            state.copy(cart = state.cart.copy(items = state.cart.items + newItem))
        }
        debounceSave()
    }

    fun incrementQuantity(productId: String) {
        _uiState.update { state ->
            val items = state.cart.items.map { item ->
                if (item.productId == productId) item.copy(quantity = item.quantity + 1) else item
            }
            state.copy(cart = state.cart.copy(items = items))
        }
        debounceSave()
    }

    fun decrementQuantity(productId: String) {
        _uiState.update { state ->
            val items = state.cart.items.mapNotNull { item ->
                when {
                    item.productId != productId -> item
                    item.quantity <= 1 -> null
                    else -> item.copy(quantity = item.quantity - 1)
                }
            }
            state.copy(cart = state.cart.copy(items = items))
        }
        debounceSave()
    }

    fun removeFromCart(index: Int) {
        _uiState.update { state ->
            val items = state.cart.items.toMutableList().apply { removeAt(index) }
            state.copy(cart = state.cart.copy(items = items))
        }
        debounceSave()
    }

    fun addSubstitute(cartItemIndex: Int, product: KrogerProduct) {
        _uiState.update { state ->
            val items = state.cart.items.toMutableList()
            if (cartItemIndex in items.indices) {
                val item = items[cartItemIndex]
                val sub = com.amoghghadge.cartquestandroid.data.model.Substitute(
                    productId = product.productId,
                    name = product.description,
                    brand = product.brand.orEmpty()
                )
                if (item.substitutes.none { it.productId == sub.productId }) {
                    items[cartItemIndex] = item.copy(substitutes = item.substitutes + sub)
                }
            }
            state.copy(cart = state.cart.copy(items = items))
        }
        debounceSave()
    }

    fun removeSubstitute(cartItemIndex: Int, substituteIndex: Int) {
        _uiState.update { state ->
            val items = state.cart.items.toMutableList()
            if (cartItemIndex in items.indices) {
                val item = items[cartItemIndex]
                val subs = item.substitutes.toMutableList()
                if (substituteIndex in subs.indices) {
                    subs.removeAt(substituteIndex)
                    items[cartItemIndex] = item.copy(substitutes = subs)
                }
            }
            state.copy(cart = state.cart.copy(items = items))
        }
        debounceSave()
    }

    fun clearCart() {
        saveJob?.cancel()
        _uiState.update { it.copy(cart = Cart(), searchQuery = "", searchResults = emptyList(), hasSearched = false) }
    }

    fun updateQuantity(index: Int, quantity: Int) {
        if (quantity < 1) {
            removeFromCart(index)
            return
        }
        _uiState.update { state ->
            val items = state.cart.items.toMutableList()
            items[index] = items[index].copy(quantity = quantity)
            state.copy(cart = state.cart.copy(items = items))
        }
        debounceSave()
    }

    suspend fun saveCartNow() {
        saveJob?.cancel()
        _uiState.update { it.copy(isSaving = true) }
        try {
            val savedCart = cartRepository.saveCart(_uiState.value.cart)
            _uiState.update { it.copy(cart = savedCart, isSaving = false) }
        } catch (_: Exception) {
            _uiState.update { it.copy(isSaving = false) }
        }
    }

    private fun debounceSave() {
        saveJob?.cancel()
        saveJob = viewModelScope.launch {
            delay(1000)
            _uiState.update { it.copy(isSaving = true) }
            try {
                val savedCart = cartRepository.saveCart(_uiState.value.cart)
                _uiState.update { it.copy(cart = savedCart, isSaving = false) }
            } catch (_: Exception) {
                _uiState.update { it.copy(isSaving = false) }
            }
        }
    }
}
